docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+power domain, etc.) as some other pods, or, conversely, should *not* be placed in the
+same topological domain as some other pods.
+
+Here are a few example rules; we explain how to express them using the API described
+in this doc later, in the section "Examples."
+* Affinity
+  * Co-locate the pods from a particular Service or Job in the same availability zone
+  * When scheduling a pod that uses a license for a particular piece of software, try
+to schedule it onto a node that is already running a pod that requires that same software,
+so that the new pod does not incur any additional license cost.
+  * Co-locate the pods from service S1 with pods from service S2 because S1 uses S2
+  and thus it is useful to minimize the network latency between them. Co-location
+  might mean same nodes and/or same availability zone.
+* Anti-affinity
+  * Spread the pods of a service across nodes and/or availability zones
+  * Give a pod "exclusive" access to a node -- it must never share the node with other pods
+  * Don't schedule the pods of a particular service on the same nodes as pods of
+  another service that are known to interfere with the performance of the pods of the first service.
+
+For both affinity and anti-affinity, there are two variants: *hard* and
+*soft.* *Hard* rules *must* be satisfied for a pod to schedule, while *soft*
+rules are *hints* that the scheduler will try to satisfy but may not be able to.
+This is similar to the difference between
+[NodeConstraints and NodePreferences](https://github.com/kubernetes/kubernetes/pull/18261).
+
+Note that this proposal is only about *inter-pod* topological affinity and anti-affinity.
+There are other forms of topological affinity and anti-affinity. For example,
+you can use NodeConstraints (NodePreferences) to require (prefer)
+that a set of pods all be scheduled in some zone Z. NodePreferences and NodeConstraints
+are not capable of expressing inter-pod dependencies, and conversely
+the API we descibe here is not capable of expressing NodePreference and NodeConstraint rules.
+For simplicity, we will use the terms "affinity" and "anti-affinity" to mean
+"inter-pod topological affinity" and "inter-pod topological anti-affinity" in the remainder
+of this document.
+
+## API
+
+We will add `AffinitySelector` to `PodSpec`.
+
+```go
+
+AffinitySelector *AffinitySelector `json:"affinitySelector,omitempty"`
+
+type AffinitySelector struct {
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAffinityExpressions []AffinityTerm `json:"hardAffinityExpression,omitempty"`
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAntiAffinityExpressions []AffinityTerm `json:"hardAntiAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
```

Comment:
See discussion in #18261 on how simple weighted sums can potentially do a much better job than expression counting.